### PR TITLE
[Mono.Profiler.Log] Support MLPD version 13

### PR DESCRIPTION
**Pull request by rodrmoya**
The differences from v14 are small, and given it is the current "stable"
MLPD version, it's worth supporting it. The differences are:

* ImagePointer field not present in v13 Assembly*Event's
* In MonitorEvent, LogMonitorEvent is not in a separate field, but included
  as a flag in the extType field.
* SampleHitEvent's have a type field in v13, which was removed in v14, because
  only cycles samples are emitted.
* MONO_GC_EVENT_{MARK,RECLAIM}_{START,END} were removed in v14
* SampleUnmanagedBinary's SegmentPointer is not based on ptr_base in v13

### CODE REVIEW

**Comment by alexrp** [comment]
This is missing:

```
changed address field in TYPE_SAMPLE_UBIN to be based on ptr_base
removed MONO_GC_EVENT_{MARK,RECLAIM}_{START,END}
```

The latter of which is one of the reasons why I didn't want to support v13. Now we have to add those enum members to `LogGCEvent` even though they don't exist in the format version that was actually 'current' when the library was introduced. 😞

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEnums.cs [hunk 1]
```diff
@@ -173,6 +173,11 @@ public enum LogGCEvent {
 		PreStartWorld = 8,
 		PostStartWorld = 9,
 		PostStartWorldUnlocked = 11,
+		// Following are v13 and older only
+		MarkStart = 1,
```
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEnums.cs]: These should use begin/end naming for consistency.

**Comment by alexrp** [comment]
You'll need to update this PR when #5562 is merged.

**Comment by rodrmoya** [comment]
Rebased with #5562

**Comment by alexrp** [comment]
I mean, you need to read the root type field as a byte for version 13.

**Comment by rodrmoya** [comment]
yes, done after the rebase, or am I missing anything else?

**Comment by alexrp** [comment]
That looks ok.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs [hunk 2]
```diff
@@ -414,10 +423,14 @@ LogEvent ReadEvent ()
...
 				case LogEventType.SampleHit:
+					if (StreamHeader.FormatVersion < 14) {
+						// Read SampleType (always set to .Cycles) for versions < 14
+						Reader.ReadByte ();
+					}
 					ev = new SampleHitEvent {
 						ThreadId = ReadPointer (),
 						UnmanagedBacktrace = ReadBacktrace (true, false),
-						ManagedBacktrace = ReadBacktrace (true),
+						ManagedBacktrace = ReadBacktrace (true).Reverse ().ToList (),
```
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: `ToArray` would be better.

Comment (rodrmoya) [reply mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogProcessor.cs]: Fixed

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs [hunk 3]
```diff
@@ -7,6 +7,8 @@
 namespace Mono.Profiler.Log {
 
 	public sealed class LogStreamHeader {
+		const int MinimumMLPDSupportedVersion = 13;
```
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs]: Let's just call these `MinVersion`/`MaxVersion`. It's obvious from the context what they refer to. Also, empty line after the `class ... {` as per the Mono coding guidelines.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs [hunk 4]
```diff
@@ -41,6 +43,10 @@ internal LogStreamHeader (LogReader reader)
 
 			Version = new Version (reader.ReadByte (), reader.ReadByte ());
 			FormatVersion = reader.ReadByte ();
+
+			if (FormatVersion < MinimumMLPDSupportedVersion || FormatVersion > MaximumMLPDSupportedVersion)
+				throw new LogException ($"Unsupported MLPD version {FormatVersion}. Should be >= {MinimumMLPDSupportedVersion} and <= {MaximumMLPDSupportedVersion}");
```
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs]: Missing period at the end of the message.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEnums.cs [hunk 5]
```diff
@@ -173,6 +173,11 @@ public enum LogGCEvent {
 		PreStartWorld = 8,
 		PostStartWorld = 9,
 		PostStartWorldUnlocked = 11,
+		// Following are v13 and older only
+		MarkBegin = 1,
```
Comment (alexrp) [inline mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEnums.cs]: We should tag these with `[Obsolete]`.

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs (hunk 3 above)
Comment (rodrmoya) [reply mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs]: done

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs (hunk 4 above)
Comment (rodrmoya) [reply mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogStreamHeader.cs]: done

File: mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEnums.cs (hunk 5 above)
Comment (rodrmoya) [reply mcs/class/Mono.Profiler.Log/Mono.Profiler.Log/LogEnums.cs]: done

**Pull request merged**